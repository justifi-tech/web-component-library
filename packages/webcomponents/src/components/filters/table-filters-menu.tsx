import { Component, Listen, Prop, State, Watch, h } from '@stencil/core';
import { Button } from '../../ui-components/button';
import { hasFilters } from './utils';
import { dropdownMenu } from '../../styles/parts';

@Component({
  tag: 'table-filters-menu'
})
export class TableFiltersMenu {
  @State() anchorButton: HTMLElement;
  @State() menu: HTMLElement;
  @State() showClearButton: boolean = false;

  @Prop() params: any;
  @Prop() clearParams: () => void;

  @Watch('params')
  watchParamsHandler(newValue: any) {
    this.showClearButton = hasFilters(newValue);
  }

  @Listen('showEvent')
  show() {
    this.menu.classList.add('show');
  }

  @Listen('hideEvent')
  hide() {
    this.menu.classList.remove('show');
  }

  componentWillLoad() {
    this.showClearButton = hasFilters(this.params);
  }

  render() {
    return (
      <div class="d-flex gap-2">
        <div class="dropdown pb-3">
          <Button
            variant="primary"
            type="button dropdown-toggle"
            ref={(el) => (this.anchorButton = el)}
            class="btn btn-primary dropdown-toggle"
            data-test-id="open-filters-button"
            data-bs-toggle="dropdown"
          >
            {"Filters"}
          </Button>
          <custom-popper
            offset={[20, 10]}
            strategy="fixed"
            anchorRef={this.anchorButton}
            trigger="click"
          >
            <form
              class="dropdown-menu"
              part={dropdownMenu}
              data-test-id="filter-menu"
              ref={(el) => (this.menu = el)}
              onSubmit={(e) => e.preventDefault()}
            >
              {/* Render filter inputs here */}
              <slot />
            </form>
          </custom-popper>
        </div>
        <div>
          <Button
            variant="secondary"
            onClick={() => this.clearParams()}
            hidden={!this.showClearButton}
            data-test-id="clear-filters-button"
          >
            {"Clear Filters"}
          </Button>
        </div>
      </div>
    );
  }
}
