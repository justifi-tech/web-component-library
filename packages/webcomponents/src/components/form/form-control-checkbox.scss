@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

// input[type='checkbox'][disabled] {
//   color: var(--jfi-form-control-disabled-color);
//   opacity: 0.5;
//   cursor: not-allowed;
// }

*::part(input[type='checkbox']) {
  margin: var(--jfi-form-control-margin);
  padding: var(--jfi-form-control-padding);
}