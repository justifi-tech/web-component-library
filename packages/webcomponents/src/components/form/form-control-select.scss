@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

select[disabled] {
  background-color: #e9ecef;
  opacity: 0.5;
  cursor: not-allowed;
}
