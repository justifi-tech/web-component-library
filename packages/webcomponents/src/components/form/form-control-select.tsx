import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-select'
})
export class SelectInput {
  selectElement!: HTMLSelectElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() options: { label: string; value: string }[];
  @Prop() disabled: boolean;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  updateInput(newValue: any) {
    this.selectElement.value = newValue;
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit(target.value);
  }

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <label part="label" class="form-label" htmlFor={this.name}>
            {this.label}
          </label>
          <select
            ref={el => (this.selectElement = el as HTMLSelectElement)}
            id={this.name}
            name={this.name}
            onInput={(event: any) => this.handleFormControlInput(event)}
            onBlur={() => this.formControlBlur.emit()}
            part={`input ${this.errorText ? 'input-invalid' : ''}`}
            class={this.errorText ? 'form-select is-invalid' : 'form-select'}
            disabled={this.disabled}
          >
            {this.options?.map(option => (
              <option value={option.value}>{option.label}</option>
            ))}
          </select>
          <form-control-help-text helpText={this.helpText} />
          <form-control-error-text errorText={this.errorText} />
        </div>
      </Host>
    );
  }
}
