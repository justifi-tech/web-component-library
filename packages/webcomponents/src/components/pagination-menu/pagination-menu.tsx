import { Component, h, Prop } from '@stencil/core';
import { ExtendedPagingDefaults, ExtendedPagingInfo } from '../../api/Pagination';
import { StyledHost } from '../../ui-components';

export interface PaginationProps {
  paging: ExtendedPagingInfo;
  params?: {};
}

@Component({
  tag: 'pagination-menu'
})

export class PaginationMenu {
  @Prop() paging: PaginationProps['paging'] = ExtendedPagingDefaults;
  @Prop() params: PaginationProps['params'] = {};

  render() {
    return (
      <StyledHost exportparts="pagination-buttons-container">
        <nav
          aria-label="Table pagination"
          class="d-flex justify-content-end gap-3"
          part="pagination-bar"
        >
          <button
            class="page-link"
            onClick={(e) => {
              e.preventDefault();
              this.paging.handleClickPrevious(this.paging.start_cursor);
            }}
            part={`button-link ${this.paging.has_next ? '' : ' button-disabled'}`}
          >
            <span class="me-1" part="page-arrow">&laquo;</span>
            <span part="page-button-text">Previous</span>
          </button>
          <button
            class="page-link"
            onClick={(e) => {
              e.preventDefault();
              this.paging.handleClickNext(this.paging.end_cursor);
            }}
            part={`button-link ${this.paging.has_next ? '' : ' button-disabled'}`}
          >
            <span part="page-button-text">Next</span>
            <span class="ms-1" part="page-arrow">&raquo;</span>
          </button>
        </nav>
      </StyledHost >
    );
  }
}
