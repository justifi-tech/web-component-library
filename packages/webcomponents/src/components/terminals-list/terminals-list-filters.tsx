import { Component, h, Prop } from '@stencil/core';
import { debounce } from 'lodash';
import { filterParams, propsParams, clearParams } from './terminals-list-params-state';
import { StyledHost } from '../../ui-components';
import { ITerminalStatus } from '../../api';

@Component({
  tag: 'justifi-terminals-list-filters',
  shadow: true
})
export class TerminalsListFilters {
  @Prop() terminalId?: string;
  @Prop() terminalStatus?: ITerminalStatus;

  private debouncedSetParamsOnChange: (name: string, value: string) => void;

  componentWillLoad() {
    // debounced input handler for text input
    this.debouncedSetParamsOnChange = debounce(this.setParamsOnChange, 300);

    const propsToSet = {
      terminal_id: this.terminalId,
      status: this.terminalStatus,
    };

    Object.entries(propsToSet).forEach(([key, value]) => {
      if (value) {
        propsParams[key] = value;
      }
    });
  }

  setParamsOnChange = (name: string, value: string) => {
    filterParams[name] = value;
  };

  get terminalStatusOptions(): { label: string, value: ITerminalStatus | '' }[] {
    return [
      { label: 'All', value: '' },
      { label: 'Connected', value: ITerminalStatus.connected },
      { label: 'Disconnected', value: ITerminalStatus.disconnected },
      { label: 'Unknown', value: ITerminalStatus.unknown },
    ]
  }

  render() {
    const filterMenuParams = { ...filterParams }

    return (
      <StyledHost>
        <table-filters-menu params={filterMenuParams} clearParams={clearParams}>
          <div class="grid-cols-2 gap-3 p-1">
            <div class="p-2">
              <form-control-text 
                name="terminal_id"
                label="Terminal ID"
                inputHandler={this.debouncedSetParamsOnChange}
                defaultValue={this.terminalId || filterParams.terminal_id}
                disabled={!!this.terminalId}
              />
            </div>
            <div class="p-2">
              <form-control-select 
                name="status"
                label="Status"
                options={this.terminalStatusOptions}
                inputHandler={this.setParamsOnChange}
                defaultValue={this.terminalStatus || filterParams.status}
                disabled={!!this.terminalStatus}
              />
            </div>
          </div>
        </table-filters-menu>
      </StyledHost>
    );
  }
}
