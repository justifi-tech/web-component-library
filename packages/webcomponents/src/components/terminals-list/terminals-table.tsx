import { h } from '@stencil/core';
import { MapTerminalStatusToBadge } from './terminals-status';

export const defaultColumnsKeys = 'nickname,provider_id,status';

export const terminalTableColumns = {
  nickname: () => (
    <th part="table-head-cell" scope="col" title="The nickname of the terminal">
      Terminal Nickname
    </th>
  ),
  provider_id: () => (
    <th part="table-head-cell" scope="col" title="The provider ID of the terminal">
      Provider ID
    </th>
  ),
  sub_account_name: () => (
    <th part="table-head-cell" scope="col" title="The sub account associated with the terminal">
      Sub Account
    </th>
  ),
  status: () => (
    <th part="table-head-cell" scope="col" title="The current status of each terminal">
      Status
    </th>
  ),
}

export const terminalTableCells = {
  nickname: (value) => (<td>{value}</td>),
  provider_id: (value) => (<td>{value}</td>),
  sub_account_name: (value) => (<td>{value}</td>),
  status: (value) => (<td innerHTML={MapTerminalStatusToBadge(value)}></td>),
}
