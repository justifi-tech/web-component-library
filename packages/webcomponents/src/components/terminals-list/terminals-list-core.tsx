import { Component, h, Prop, State, Watch, Event, EventEmitter } from '@stencil/core';
import { PagingInfo, SubAccount, Terminal, pagingDefaults } from '../../api';
import { TableEmptyState, TableErrorState, TableLoadingState } from '../../ui-components';
import { terminalTableColumns, terminalTableCells } from './terminals-table';
import { Table } from '../../utils/table';
import { getRequestParams, onQueryParamsChange } from './terminals-list-params-state';
import { ComponentClickEvent, ComponentErrorEvent } from '../../api/ComponentEvents';
import {
  TableWrapper,
  TableComponent,
  TableHead,
  TableHeadRow,
  TableBody,
  TableRow,
  TableFoot,
  TableFootRow,
  TableFootCell,
  TableClickActions
} from '../../ui-components';

@Component({
  tag: 'terminals-list-core'
})
export class TerminalsListCore {
  @Prop() getTerminals: Function;
  @Prop() getSubAccounts: Function;
  @Prop() columns: string;

  @State() terminals: Terminal[] = [];
  @State() terminalsTable: Table<Terminal>;
  @State() subAccounts: SubAccount[] = [];
  @State() loading: boolean = true;
  @State() errorMessage: string;
  @State() paging: PagingInfo = pagingDefaults;
  @State() pagingParams: any = {};
  
  @Watch('pagingParams')
  @Watch('getTerminals')
  @Watch('getSubAccounts')
  @Watch('columns')
  updateOnPropChange() {
    this.fetchData();
  }

  @Event({ eventName: 'click-event', bubbles: true }) clickEvent: EventEmitter<ComponentClickEvent>;
  @Event({ eventName: 'error-event' }) errorEvent: EventEmitter<ComponentErrorEvent>;

  componentWillLoad() {
    this.terminalsTable = new Table<Terminal>(this.terminals, this.columns, terminalTableColumns, terminalTableCells);
    if (this.getTerminals && this.getSubAccounts) {
      this.fetchData();
    }

    onQueryParamsChange('set', () => {
      this.pagingParams = {};
    });

    onQueryParamsChange('reset', () => {
      this.pagingParams = {};
      this.errorMessage = '';
    });
  }

  fetchData(): void {
    this.loading = true;

    this.getTerminals({
      params: this.terminalParams,
      onSuccess: async ({ terminals, pagingInfo }) => {
        this.terminals = terminals;
        this.paging = pagingInfo;
        this.terminalsTable.collectionData = this.terminals;
        const shouldFetchSubAccounts = this.terminalsTable.columnKeys.includes('sub_account_name');

        if (shouldFetchSubAccounts) {
          await this.fetchSubAccounts();
        } else {
          this.loading = false;
        }

      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  async fetchSubAccounts(): Promise<void> {
    this.getSubAccounts({
      params: this.subAccountParams,
      onSuccess: ({ subAccounts }) => {
        this.subAccounts = subAccounts;
        this.terminals = this.terminals.map((terminal) => {
          terminal.sub_account_name = this.subAccounts.find((subAccount) => subAccount.id === terminal.account_id)?.name;
          return terminal;
        });
        this.loading = false;
      },
      onError: ({ error, code, severity }) => {
        this.errorMessage = error;
        this.errorEvent.emit({
          errorCode: code,
          message: error,
          severity,
        });
        this.loading = false;
      },
    });
  }

  handleClickPrevious = (beforeCursor: string) => {
    this.pagingParams = { before_cursor: beforeCursor };
    this.clickEvent.emit({ name: TableClickActions.previous });
  };

  handleClickNext = (afterCursor: string) => {
    this.pagingParams = { after_cursor: afterCursor };
    this.clickEvent.emit({ name: TableClickActions.next });
  };

  rowClickHandler = (e) => {
    const clickedTerminalID = e.target.closest('tr').dataset.rowEntityId;
    if (!clickedTerminalID) return;

    const terminalData = this.terminals.find((terminal) => terminal.id === clickedTerminalID);
    this.clickEvent.emit({ name: TableClickActions.row, data: terminalData });
  };

  get terminalParams() {
    const requestParams = getRequestParams();
    const params = { ...requestParams, ...this.pagingParams };
    return params;
  }

  get subAccountParams() {
    let accountIdNumbers = this.terminals.map((terminal) => terminal.account_id);
    let uniqueAccountIds = [...new Set(accountIdNumbers)];
    let accountIdString = uniqueAccountIds.join(',');
    return { sub_account_id: accountIdString };
  }

  get entityId() {
    return this.terminals.map((terminal) => terminal.id);
  }

  get showEmptyState() {
    return !this.loading && !this.errorMessage && this.terminalsTable.rowData.length < 1;
  }

  get showErrorState() {
    return !this.loading && !!this.errorMessage;
  }

  get showRowData() {
    return !this.showEmptyState && !this.showErrorState && !this.loading;
  }

  render() {
    return (
      <TableWrapper>
        <TableComponent>
          <TableHead>
            <TableHeadRow>
              {this.terminalsTable.columnData.map((column) => column)}
            </TableHeadRow>
          </TableHead>
          <TableBody>
            <TableLoadingState
              columnSpan={this.terminalsTable.columnKeys.length}
              isLoading={this.loading}
            />
            <TableEmptyState
              isEmpty={this.showEmptyState}
              columnSpan={this.terminalsTable.columnKeys.length}
            />
            <TableErrorState
              columnSpan={this.terminalsTable.columnKeys.length}
              errorMessage={this.errorMessage}
            />
            {this.showRowData && this.terminalsTable.rowData.map((data, index) => (
              <TableRow
                data-test-id="table-row"
                data-row-entity-id={this.entityId[index]}
                onClick={this.rowClickHandler}
              >
                {data}
              </TableRow>
            ))}
          </TableBody>
          {this.paging && (
            <TableFoot>
              <TableFootRow>
                <TableFootCell colSpan={this.terminalsTable.columnData.length}>
                  <pagination-menu
                    paging={{
                      ...this.paging,
                      handleClickPrevious: this.handleClickPrevious,
                      handleClickNext: this.handleClickNext,
                    }}
                  />
                </TableFootCell>
              </TableFootRow>
            </TableFoot>
          )}
        </TableComponent>
      </TableWrapper>
    );
  }
}
