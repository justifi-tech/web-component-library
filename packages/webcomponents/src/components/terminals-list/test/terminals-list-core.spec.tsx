jest.mock('../../../ui-components/styled-host/styled-host.css', () => '');

import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { TerminalsListCore } from '../terminals-list-core';
import { PaginationMenu } from '../../pagination-menu/pagination-menu';
import mockTerminalSuccessResponse from '../../../../../../mockData/mockTerminalsListSuccess.json';
import mockSubAccountSuccessResponse from '../../../../../../mockData/mockSubAccountsListSuccess.json';
import { IApiResponseCollection, ISubAccount, ITerminal } from '../../../api';
import { makeGetTerminals } from '../get-terminals';
import { TerminalsListFilters } from '../terminals-list-filters';
import { TableFiltersMenu } from '../../../ui-components/filters/table-filters-menu';
import { SelectInput } from '../../../ui-components/form/form-control-select';
import { makeGetSubAccounts } from '../../../api/get-subaccounts';
import { defaultColumnsKeys } from '../terminals-table';

const mockTerminalsResponse = mockTerminalSuccessResponse as IApiResponseCollection<ITerminal[]>;
const mockSubAccountsResponse = mockSubAccountSuccessResponse as IApiResponseCollection<ISubAccount[]>;
const components = [TerminalsListCore, PaginationMenu, TableFiltersMenu, TerminalsListFilters, SelectInput];

describe('terminals-list-core', () => {
  it('renders properly with fetched data', async () => {
    const mockTerminalsService = {
      fetchTerminals: jest.fn().mockResolvedValue(mockTerminalsResponse),
    };

    const getTerminals = makeGetTerminals({
      id: '123',
      authToken: '123',
      service: mockTerminalsService,
      apiOrigin: 'http://localhost:3000'
    });
    
    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    expect(page.rootInstance.terminals[0]).toEqual(expect.objectContaining({ account_id: mockTerminalsResponse.data[0].account_id }));
    const rows = page.root.querySelectorAll('[data-test-id="table-row"]');
    expect(rows.length).toBe(mockTerminalsResponse.data.length);
    expect(mockTerminalsService.fetchTerminals).toHaveBeenCalled();
    expect(page.root).toMatchSnapshot();
  });

  it('displays an error state on failed data fetch', async () => {
    const mockService = {
      fetchTerminals: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getTerminals = makeGetTerminals({
      id: 'some-id',
      authToken: 'some-auth-token',
      service: mockService,
      apiOrigin: 'http://localhost:3000'
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    expect(page.rootInstance.errorMessage).toBe('Fetch error');
    expect(page.root).toMatchSnapshot();
  });

  it('emits row-clicked event on row click', async () => {
    const mockTerminalsService = {
      fetchTerminals: jest.fn().mockResolvedValue(mockTerminalsResponse),
    };

    const getTerminals = makeGetTerminals({
      id: '123',
      authToken: '123',
      service: mockTerminalsService,
      apiOrigin: 'http://localhost:3000'
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    const firstRow = page.root.querySelector('[data-test-id="table-row"]') as HTMLElement;
    expect(firstRow).not.toBeNull();

    const spyEvent = jest.fn();
    page.win.addEventListener('row-clicked', spyEvent);

    firstRow.click();
    expect(spyEvent).toHaveBeenCalled();
  });

  it('shows table filter menu on filter button click', async () => {
    const mockTerminalsService = {
      fetchTerminals: jest.fn().mockResolvedValue(mockTerminalsResponse),
    };

    const getTerminals = makeGetTerminals({
      id: '123',
      authToken: '123',
      service: mockTerminalsService,
      apiOrigin: 'http://localhost:3000'
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    const filterButton = page.root.querySelector('[data-test-id="open-filters-button"]') as HTMLElement;
    expect(filterButton).not.toBeNull();

    filterButton.click();
    await page.waitForChanges();

    const filterMenu = page.root.querySelector('[data-test-id="filter-menu"]') as HTMLElement;
    expect(filterMenu).not.toBeNull();
  });

  it('updates params and refetches data on filter interaction', async () => {
    const mockTerminalsService = {
      fetchTerminals: jest.fn().mockResolvedValue(mockTerminalsResponse),
    };

    const getTerminals = makeGetTerminals({
      id: '123',
      authToken: '123',
      service: mockTerminalsService,
      apiOrigin: 'http://localhost:3000'
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    const filterButton = page.root.querySelector('[data-test-id="open-filters-button"]') as HTMLElement;
    filterButton.click();

    const filterMenu = page.root.querySelector('terminals-list-filters') as HTMLElement;
    expect(filterMenu).not.toBeNull();

    const selectFilter = filterMenu.querySelector('form-control-select') as HTMLFormControlSelectElement;
    expect(selectFilter).not.toBeNull();

    const selectFilterInput = selectFilter.querySelector('select') as HTMLSelectElement;
    expect(selectFilterInput).not.toBeNull();

    selectFilterInput.click();
    
    const selectOptions = selectFilterInput.querySelectorAll('option');
    expect(selectOptions).not.toBeNull();
    
    const succeededOption = selectOptions[3] as HTMLOptionElement;
    succeededOption.click();
    selectFilterInput.value = 'connected';
    selectFilterInput.dispatchEvent(new Event('input'));

    expect(mockTerminalsService.fetchTerminals).toHaveBeenCalledTimes(2);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams).toEqual({"status": "connected"});
  });

  it('clears filters and refetches data on clear filters interaction', async () => {
    const mockTerminalsService = {
      fetchTerminals: jest.fn().mockResolvedValue(mockTerminalsResponse),
    };

    const getTerminals = makeGetTerminals({
      id: '123',
      authToken: '123',
      service: mockTerminalsService,
      apiOrigin: 'http://localhost:3000'
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    const filterButton = page.root.querySelector('[data-test-id="open-filters-button"]') as HTMLElement;
    filterButton.click();

    const filterMenu = page.root.querySelector('table-filters-menu') as HTMLElement;
    expect(filterMenu).not.toBeNull();

    const selectFilter = filterMenu.querySelector('form-control-select') as HTMLFormControlSelectElement;
    expect(selectFilter).not.toBeNull();

    const selectFilterInput = selectFilter.querySelector('select') as HTMLSelectElement;
    expect(selectFilterInput).not.toBeNull();

    selectFilterInput.click();
    
    const selectOptions = selectFilterInput.querySelectorAll('option');
    expect(selectOptions).not.toBeNull();
    
    const succeededOption = selectOptions[3] as HTMLOptionElement;
    succeededOption.click();
    selectFilterInput.value = 'connected';
    selectFilterInput.dispatchEvent(new Event('input'));

    const clearButton = filterMenu.querySelector('[data-test-id="clear-filters-button"]') as HTMLElement;
    expect(clearButton).not.toBeNull();

    clearButton.click();

    expect(mockTerminalsService.fetchTerminals).toHaveBeenCalledTimes(3);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams).toEqual({});
  });

  it('updates params and refetches data on pagination interaction', async () => {
    const mockTerminalsService = {
      fetchTerminals: jest.fn().mockResolvedValue(mockTerminalsResponse),
    };

    const getTerminals = makeGetTerminals({
      id: '123',
      authToken: '123',
      service: mockTerminalsService,
      apiOrigin: 'http://localhost:3000'
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} />,
    });

    await page.waitForChanges();

    // Assuming handleClickNext is accessible and modifies `params` which triggers a re-fetch
    page.rootInstance.handleClickNext('nextCursor');
    await page.waitForChanges();

    // The mock function should be called 2 times: once for the initial load and later after the pagination interaction
    expect(mockTerminalsService.fetchTerminals).toHaveBeenCalledTimes(2);
    const updatedParams = page.rootInstance.params;
    expect(updatedParams.after_cursor).toBe('nextCursor');
  });

  it('emits error event on fetch error', async () => {
    const mockService = {
      fetchTerminals: jest.fn().mockRejectedValue(new Error('Fetch error'))
    };

    const getTerminals = makeGetTerminals({
      id: 'some-id',
      authToken: 'some-auth',
      service: mockService,
      apiOrigin: 'http://localhost:3000'
    });

    const mockSubAccountsService = {
      fetchSubAccounts: jest.fn().mockResolvedValue(mockSubAccountsResponse),
    };

    const getSubAccounts = makeGetSubAccounts({
      accountId: '123',
      authToken: '123',
      service: mockSubAccountsService,
      apiOrigin: 'http://localhost:3000'
    });

    const errorEvent = jest.fn();

    const page = await newSpecPage({
      components: components,
      template: () => <terminals-list-core getTerminals={getTerminals} getSubAccounts={getSubAccounts} columns={defaultColumnsKeys} onError-event={errorEvent} />,
    });

    await page.waitForChanges();

    expect(errorEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        detail: {
          errorCode: 'fetch-error',
          message: 'Fetch error',
          severity: 'error',
        }
      })
    );
  });
});
