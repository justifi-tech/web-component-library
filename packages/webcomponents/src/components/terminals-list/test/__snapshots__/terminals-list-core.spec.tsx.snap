// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terminals-list-core displays an error state on failed data fetch 1`] = `
<terminals-list-core>
  <div>
    <div class="table-wrapper">
      <table class="table table-hover" part="table text color font-family background-color">
        <thead class="sticky-top table-head">
          <tr class="table-light text-nowrap">
            <th part="table-head-cell table-cell text color font-family background-color text color font-family background-color" scope="col" title="The nickname associated with the terminal">
              Nickname
            </th>
            <th part="table-head-cell table-cell text color font-family background-color text color font-family background-color" scope="col" title="The serial number / provider ID of the terminal">
              Serial Number
            </th>
            <th part="table-head-cell table-cell text color font-family background-color text color font-family background-color" scope="col" title="The current status of each terminal">
              Status
            </th>
          </tr>
        </thead>
        <tbody class="table-body">
          <tr>
            <td class="error-state" colspan="3" data-test-id="table-error-state" part="table-error text color font-family background-color" style="text-align: center;">
              An unexpected error occurred: Fetch error
            </td>
          </tr>
        </tbody>
        <tfoot class="sticky-bottom">
          <tr class="align-middle table-light">
            <td colspan="3" part="table-cell text color font-family background-color">
              <pagination-menu>
                <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                  <ul class="m-0 pagination">
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="button-disabled button text color font-family">
                        <span class="me-1">
                          «
                        </span>
                        <span>
                          Previous
                        </span>
                      </a>
                    </li>
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="button-disabled button text color font-family">
                        <span>
                          Next
                        </span>
                        <span class="ms-1">
                          »
                        </span>
                      </a>
                    </li>
                  </ul>
                </nav>
              </pagination-menu>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</terminals-list-core>
`;

exports[`terminals-list-core renders properly with fetched data 1`] = `
<terminals-list-core>
  <div>
    <div class="table-wrapper">
      <table class="table table-hover" part="table text color font-family background-color">
        <thead class="sticky-top table-head">
          <tr class="table-light text-nowrap">
            <th part="table-head-cell table-cell text color font-family background-color text color font-family background-color" scope="col" title="The nickname associated with the terminal">
              Nickname
            </th>
            <th part="table-head-cell table-cell text color font-family background-color text color font-family background-color" scope="col" title="The serial number / provider ID of the terminal">
              Serial Number
            </th>
            <th part="table-head-cell table-cell text color font-family background-color text color font-family background-color" scope="col" title="The current status of each terminal">
              Status
            </th>
          </tr>
        </thead>
        <tbody class="table-body">
          <tr data-row-entity-id="trm_43Y2lE9E2ULR8pvaKI6D5v" data-test-id="table-row">
            <td part="table-cell-odd table-cell text color font-family background-color text color font-family background-color">
              Terminal 1
            </td>
            <td part="table-cell-odd table-cell text color font-family background-color text color font-family background-color">
              453-760-515
            </td>
            <td part="table-cell-odd table-cell text color font-family background-color text color font-family background-color">
              <span class="badge bg-danger" part="badge font-family badge-danger" title="This terminal is disconnected">
                Disconnected
              </span>
            </td>
          </tr>
          <tr data-row-entity-id="trm_6TFOXK5U0LZLyQ6y6t3sbj" data-test-id="table-row">
            <td part="table-cell-even table-cell text color font-family background-color text color font-family background-color">
              Terminal 2
            </td>
            <td part="table-cell-even table-cell text color font-family background-color text color font-family background-color">
              807-075-625
            </td>
            <td part="table-cell-even table-cell text color font-family background-color text color font-family background-color">
              <span class="badge bg-danger" part="badge font-family badge-danger" title="This terminal is disconnected">
                Disconnected
              </span>
            </td>
          </tr>
          <tr data-row-entity-id="trm_F8UyFyXcaTBwSDh1cMxjc" data-test-id="table-row">
            <td part="table-cell-odd table-cell text color font-family background-color text color font-family background-color">
              Terminal 3
            </td>
            <td part="table-cell-odd table-cell text color font-family background-color text color font-family background-color">
              450-060-752
            </td>
            <td part="table-cell-odd table-cell text color font-family background-color text color font-family background-color">
              <span class="badge bg-success" part="badge font-family badge-success" title="This terminal is connected">
                Connected
              </span>
            </td>
          </tr>
        </tbody>
        <tfoot class="sticky-bottom">
          <tr class="align-middle table-light">
            <td colspan="3" part="table-cell text color font-family background-color">
              <pagination-menu>
                <nav aria-label="Table pagination" class="d-flex gap-3 justify-content-end">
                  <ul class="m-0 pagination">
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="button-disabled button text color font-family">
                        <span class="me-1">
                          «
                        </span>
                        <span>
                          Previous
                        </span>
                      </a>
                    </li>
                    <li class="disabled page-item">
                      <a class="page-link" href="#" part="button-disabled button text color font-family">
                        <span>
                          Next
                        </span>
                        <span class="ms-1">
                          »
                        </span>
                      </a>
                    </li>
                  </ul>
                </nav>
              </pagination-menu>
            </td>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</terminals-list-core>
`;
